src/app/static/src/app/utils.ts

```diff
@@ -162,6 +162,35 @@ const flattenOption = (filterOption: NestedFilterOption): NestedFilterOption =>
     return result;
 };
 
+export const flattenOptionsIdsByHierarchy = (filterOptions: NestedFilterOption[]): string[] => {
+    const result: any[] = [];
+    let layer = 0
+
+    function recursive(filterOptions: any[]){
+        const filterOptionArray = filterOptions.map((option: NestedFilterOption) => option.id)
+        if (result.length > layer){
+            result[layer] = [...result[layer],  ...filterOptionArray]
+        } else {
+            result.push(filterOptionArray)
+        }
+        const parentLayer = layer
+        filterOptions.forEach(filterOption => {
+            layer = parentLayer
+            if (filterOption.children?.length){
+                layer += 1
+                recursive(filterOption.children)
+            }
+        });
+    }
+    recursive(filterOptions)
+    
+    let output: string[] = []
+    result.forEach(r => {
+        output = [...output, ...r]
+    })
+    return output;
+};
+
```

Comment:
This works nicely, but I think it could be simplified a bit. What you're doing here is building up the `result` array where each item is a list of options at each layer, but I don't think this is necessary - we  should be able build up the output id array directly by recursing at each layer rather than for the children of each filterOption (so what gets passed to the next call of `recursive` is the concatenated list of filter options at a layer).  Something like this?

```suggestion
export const flattenOptionsIdsByHierarchy = (filterOptions: NestedFilterOption[]): string[] => {
    const result: string[] = [];
    const recursive = (filterOptions: NestedFilterOption[]) => {
        // 1. Push ids for the top level of options
        filterOptions.forEach((option: NestedFilterOption) => result.push(option.id));
        let nextLayer: NestedFilterOption[] = [];
        // 2. Get all options at next layer and recurse
        filterOptions.forEach((option: NestedFilterOption) => {
            if (option.children) {
                nextLayer = nextLayer.concat(option.children as NestedFilterOption[]);
            }
        });
        if (nextLayer.length > 0) {
            recursive(nextLayer);
        }
    };
    recursive(filterOptions);
    return result;
};

```

Reply:
> This works nicely, but I think it could be simplified a bit. What you're doing here is building up the `result` array where each item is a list of options at each layer, but I don't think this is necessary - we should be able build up the output id array directly by recursing at each layer rather than for the children of each filterOption (so what gets passed to the next call of `recursive` is the concatenated list of filter options at a layer). Something like this?

Ok great. Have you tested this? I had something similar to this originally, but I found as I went down to the deeper layers (great-grand children, etc), the recursive function would start to split, resulting in great-grandchildren coming before children.

Reply:
I tested it a bit, seemed to work ok. I think this should keep everything at a layer all together - but do try to break it!

Reply:
Yeah, seems to work fine. So as I understand, you are flattening out the children of layers below into a single array and then recursing, avoiding the splitting problem I mentioned.